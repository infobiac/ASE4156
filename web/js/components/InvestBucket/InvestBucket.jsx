// @flow
import React from 'react';
import PropTypes from 'prop-types';
import TextField from 'material-ui/TextField';
import Card, { CardHeader, CardContent, CardActions } from 'material-ui/Card';
import List, { ListItem, ListItemIcon, ListItemText } from 'material-ui/List';
import TrendingUpIcon from 'material-ui-icons/TrendingUp';
import TrendingDownIcon from 'material-ui-icons/TrendingDown';
import Button from 'material-ui/Button';
import AddIcon from 'material-ui-icons/Add';
import DeleteIcon from 'material-ui-icons/DeleteForever';

import type { Node } from 'react';

const propAttrShape = PropTypes.shape({ shortDesc: PropTypes.string.isRequired });

type State = {
  editMode: bool,
  editText: string,
  editGood: bool,
}
type ItemObj = {
  id: string,
  shortDesc: string,
  icon?: any,
  text?: any,
  editMode: bool,
}
type Props = {
  title: Node,
  attributes: {
    good: Array<ItemObj>,
    bad: Array<ItemObj>,
  },
  editFunc?: ?(text: string, isGood: bool) => void,
  seeMoreFunc?: ?() => void,
  editCompositionFunc?: ?() => void,
  deleteFunc?: ?() => void,
  investFunc: () => void,
}

class InvestBucket extends React.Component <Props, State> {
  static propTypes = {
    title: PropTypes.node.isRequired,
    attributes: PropTypes.shape({
      good: PropTypes.arrayOf(propAttrShape.isRequired),
      bad: PropTypes.arrayOf(propAttrShape.isRequired),
    }),
    editFunc: PropTypes.func,
    seeMoreFunc: PropTypes.func,
  }
  static defaultProps = {
    attributes: {
      good: [],
      bad: [],
    },
    editFunc: null,
    seeMoreFunc: null,
  }
  static renderAttr(attr : ItemObj, isGood : bool) {
    let indicator = null;
    if (attr.editMode) {
      indicator = <DeleteIcon id="delete-ico" />;
    } else if (isGood) {
      indicator = <TrendingUpIcon />;
    } else {
      indicator = <TrendingDownIcon />;
    }
    let text = null;
    if (attr.editMode) {
      text = <TextField id="edit-text-field" value={attr.shortDesc} {...attr.text} />;
    } else {
      text = (<ListItemText
        key={attr.id}
        primary={attr.shortDesc}
        autoFocus
        id="attr"
        {...attr.text}
      />);
    }
    return (
      <ListItem key={attr.id}>
        <ListItemIcon {...attr.icon}>
          {indicator}
        </ListItemIcon>
        {text}
      </ListItem>
    );
  }
  constructor() {
    super();
    this.state = {
      editMode: false,
      editText: '',
      editGood: true,
    };
  }
  onEnterPress = (e: SyntheticEvent<>) => {
    if (e.charCode === 13 && this.props.editFunc) {
      this.props.editFunc(this.state.editText, this.state.editGood);
      this.setState(() => ({
        editMode: false,
        editText: '',
        editGood: true,
      }));
    }
  }
  updateText = (e: SyntheticInputEvent<>) => {
    const text = e.target.value;
    this.setState(() => ({
      editText: text,
    }));
  }
  launchEdit = (mode: bool) => () => this.setState(() => ({ editMode: mode }))
  editField = () => (
    <ListItem>
      <ListItemIcon id="good-or-bad" onClick={() => this.setState(state => ({ editGood: !state.editGood }))}>
        {this.state.editGood ? <TrendingUpIcon /> : <TrendingDownIcon />}
      </ListItemIcon>
      <TextField
        autoFocus
        margin="dense"
        id="name"
        label="New Attribute"
        type="text"
        value={this.state.editText}
        onChange={this.updateText}
        onKeyPress={this.onEnterPress}
        fullWidth
      />
    </ListItem>
  )
  render() {
    return (
      <Card>
        <CardHeader title={this.props.title} />
        <CardContent>
          <List>
            {
              this.props.attributes.good.map(g => InvestBucket.renderAttr(g, true))
              .concat(this.props.attributes.bad.map(b => InvestBucket.renderAttr(b, false)))
            }
            {
              this.props.seeMoreFunc ?
                <ListItem key="seeMore">
                  <Button id="more" onClick={this.props.seeMoreFunc}>More</Button>
                </ListItem>
              : null
            }
            {
              this.state.editMode ?
                this.editField()
              : null
            }
          </List>
          {
            this.props.editFunc && !this.state.editMode ? (
              <Button fab id="launch-edit" color="primary" aria-label="add" onClick={this.launchEdit(true)}>
                <AddIcon />
              </Button>
            ) : null
          }
        </CardContent>
        <CardActions>
          <Button id="invest" dense color="primary" onClick={this.props.investFunc}>
            Invest
          </Button>
          {
            this.props.deleteFunc ? (
              <Button dense id="delete" color="primary" onClick={this.props.deleteFunc}>
                Delete
              </Button>
            ) : null
          }
          {
            this.props.editCompositionFunc ? (
              <Button dense id="edit-comp" color="primary" onClick={this.props.editCompositionFunc}>
                Composition
              </Button>
            ) : null
          }
        </CardActions>
      </Card>
    );
  }
}

export default InvestBucket;
