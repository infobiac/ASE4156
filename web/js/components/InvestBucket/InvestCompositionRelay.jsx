// @flow
import React from 'react';
import PropTypes from 'prop-types';
import { createRefetchContainer, graphql } from 'react-relay';
import { routerShape } from 'found';

import type { RelayContext } from 'react-relay';

import InvestComposition from './InvestComposition';
import changeBucketComposition from '../../mutations/BucketEdit/ChangeBucketComposition';

import type { InvestCompositionRelay_bucket } from './__generated__/InvestCompositionRelay_bucket.graphql';
import type { InvestCompositionRelay_profile } from './__generated__/InvestCompositionRelay_profile.graphql';

type ChunkList = Array<{
  id: string,
  name: string,
  quantity: number,
  value: number,
}>;

type Props = {
  bucket: InvestCompositionRelay_bucket,
  profile: InvestCompositionRelay_profile,
  relay: RelayContext,
}
type State = {
  chunks: ChunkList,
}

class InvestCompositionRelay extends React.Component<Props, State> {
  static contextTypes = {
    errorDisplay: PropTypes.func.isRequired,
    router: routerShape.isRequired,
  };
  constructor(props) {
    super();
    this.state = {
      chunks: props.bucket.stocks.edges.map(s => (s && s.node && s.node.stock.latestQuote ? ({
        id: s.node.stock.id,
        quantity: s.node.quantity,
        value: s.node.stock.latestQuote.value,
        name: s.node.stock.name,
      }) : null)).filter(x => !!x),
    };
  }
  saveChunks = (chunks) => {
    const updater = (store) => {
      const data = store.getRootField('editConfiguration').getLinkedRecord('bucket');
      store.getRoot().copyFieldsFrom(data);
      store.get(this.props.bucket.id).setValue(data.getValue('available'), 'available');
      store.get(this.props.bucket.id).setLinkedRecord(data.getLinkedRecord('stocks'), 'stocks');
    };
    changeBucketComposition(updater, null, (r, e) => this.context.errorDisplay({
      message: e ? e[0].message : 'Composition successfully changed',
    }))(this.props.relay.environment)({
      config: chunks.map(c => ({ idValue: c.id, quantity: c.quantity })),
      id: this.props.bucket.id,
    });
  }
  save = () => {
    this.saveChunks(this.state.chunks);
    this.cancel();
  }
  cancel = () => {
    this.context.router.replace('/home');
  }
  updateChunks = chunks => this.setState(() => ({ chunks }))
  render() {
    if (
      !this.props.profile.investSearch ||
      !this.props.bucket.stocks
    ) {
      return null;
    }
    return (
      <InvestComposition
        chunks={this.state.chunks}
        total={
          this.props.bucket.available
          + this.props.bucket.stocks.edges.reduce((sum, item) =>
              sum
            + (
              item && item.node && item.node.stock.latestQuote
                ? item.node.quantity * item.node.stock.latestQuote.value
                : 0
            ), 0)
        }
        chunkUpdate={this.updateChunks}
        suggestionFieldChange={(text) => {
          this.props.relay.refetch(() => ({ text }));
        }}
        suggestions={
          // $FlowFixMe
          this.props.profile.investSearch.map(s => ({
            value: s && s.latestQuote ? s.latestQuote.value : 0,
            ...s,
          }))
        }
        saveFunc={this.save}
        cancelFunc={this.cancel}
      />
    );
  }
}

export default createRefetchContainer(InvestCompositionRelay, {
  bucket: graphql`
    fragment InvestCompositionRelay_bucket on GInvestmentBucket {
      id
      available
      stocks {
        edges {
          node {
            quantity
            stock {
              id
              name
              latestQuote {
                value
              }
            }
          }
        }
      }
    }
  `,
  profile: graphql`
    fragment InvestCompositionRelay_profile on GProfile
    @argumentDefinitions(
      text: {type: "String!", defaultValue: ""}
    ) {
      investSearch: stockFind(text: $text, first: 4) {
        name
        id
        latestQuote {
          value
        }
      }
    }
  `,
}, graphql`
    query InvestCompositionRelayQuery($text: String!) {
      viewer {
        profile {
          ...InvestCompositionRelay_profile @arguments(text: $text)
        }
      }
    }
  `);
