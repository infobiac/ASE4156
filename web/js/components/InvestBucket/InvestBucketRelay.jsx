// @flow
import React from 'react';
import PropTypes from 'prop-types';
import { routerShape } from 'found';
import { createRefetchContainer, graphql } from 'react-relay';
import { ConnectionHandler } from 'relay-runtime';
import LockIcon from 'material-ui-icons/Lock';
import Dialog, {
  DialogContent,
  DialogTitle,
  DialogActions,
  DialogContentText,
} from 'material-ui/Dialog';
import Button from 'material-ui/Button';

import type { RelayContext } from 'react-relay';

import InvestBucket from './InvestBucket';
import addDescription from '../../mutations/BucketEdit/AddDescription';
import editDescription from '../../mutations/BucketEdit/EditDescription';
import deleteDescription from '../../mutations/BucketEdit/DeleteDescription';
import deleteBucket from '../../mutations/BucketEdit/DeleteBucket';

import type { InvestBucketRelay_bucket } from './__generated__/InvestBucketRelay_bucket.graphql';

type EditObj = {
  shortDesc: string,
}
type State = {
  itemCount: number,
  editMode: ?string,
  editState: EditObj,
  deleteConfirm: bool,
}
type Props = {
  bucket: InvestBucketRelay_bucket,
  relay: RelayContext,
}

class InvestBucketRelay extends React.Component<Props, State> {
  static contextTypes = {
    errorDisplay: PropTypes.func.isRequired,
    router: routerShape.isRequired,
  };
  constructor() {
    super();
    this.state = {
      itemCount: 2,
      editMode: null,
      editState: { shortDesc: '' },
      deleteConfirm: false,
    };
  }
  launchEdit = id => () => {
    this.setState((state, props) => {
      if (!props.bucket.description || !props.bucket.description.edges) {
        return {};
      }
      const item = props.bucket.description.edges.find(x => x && x.node && x.node.id === id);
      if (!item || !item.node || !item.node.text) {
        return {};
      }
      return {
        editMode: id,
        editState: {
          shortDesc: item.node.text,
        },
      };
    });
  }
  deleteBucket = () => {
    const updater = (store) => {
      store.delete(this.props.bucket.id);
    };
    deleteBucket(updater, updater, (r, error) => {
      if (error) {
        this.context.errorDisplay({
          message: error[0].message,
        });
      }
    })(this.props.relay.environment)({ id: this.props.bucket.id });
  }
  render() {
    let data;
    if (!this.props.bucket.description) {
      data = [];
    } else {
      data = this.props.bucket.description.edges;
    }
    const attributes = data.reduce((all, item) => {
      if (!item || !item.node) {
        return all;
      }
      const textAttr = {};
      const iconAttr = {};
      let extra = {};
      if (!item || !item.node || !item.node.id) {
        return all;
      }
      const { id } = item.node;
      if (this.props.bucket.isOwner) {
        if (id === this.state.editMode) {
          textAttr.onKeyPress = (e) => {
            if (e.charCode === 13) {
              this.setState(() => ({
                editMode: null,
              }), () => {
                editDescription(null, null, (r, error) => {
                  if (error) {
                    this.context.errorDisplay({
                      message: error[0].message,
                    });
                  }
                })(this.props.relay.environment)({ text: this.state.editState.shortDesc, id });
              });
            }
          };
          textAttr.onChange = (e: SyntheticInputEvent<>) => {
            const text = e.target.value;
            this.setState(() => ({
              editState: {
                shortDesc: text,
              },
            }));
          };
          textAttr.autoFocus = true;
          iconAttr.onClick = () => {
            this.setState(() => ({
              editMode: null,
            }), () => {
              const updater = (store) => {
                store.delete(id);
              };
              deleteDescription(updater, updater, (r, error) => {
                if (error) {
                  this.context.errorDisplay({
                    message: error[0].message,
                  });
                }
              })(this.props.relay.environment)({ id });
            });
          };
          extra = this.state.editState;
        } else {
          const edit = this.launchEdit(id);
          textAttr.onClick = edit;
          iconAttr.onClick = edit;
        }
      }
      all[item.node.isGood ? 'good' : 'bad'].push({
        ...item.node,
        text: textAttr,
        icon: iconAttr,
        editMode: (id === this.state.editMode),
        shortDesc: item.node.text,
        ...extra,
      });
      return all;
    }, {
      good: [],
      bad: [],
    });
    let editFunc = null;
    if (this.props.bucket.isOwner) {
      const updater = (store) => {
        const connection = ConnectionHandler.getConnection(
          store.get(this.props.bucket.id),
          'InvestBucketRelay_description',
        );
        const newEdge = ConnectionHandler.createEdge(
          store,
          connection,
          store.getRootField('addAttributeToBucket').getLinkedRecord('bucketAttr'),
          'GInvestmentBucketAttributeConnection',
        );
        ConnectionHandler.insertEdgeAfter(connection, newEdge);
      };
      editFunc = (text, isGood) => addDescription(updater, updater, (r, error) => {
        if (error) {
          this.context.errorDisplay({
            message: error[0].message,
          });
        }
      })(this.props.relay.environment)({ text, bucketId: this.props.bucket.id, isGood });
    }
    let seeMoreFunc = null;
    if (this.props.bucket.description && this.props.bucket.description.pageInfo.hasNextPage) {
      seeMoreFunc = () => {
        this.setState(
          state => ({ itemCount: state.itemCount + 2 }),
          () => this.props.relay.refetch(() => ({
            id: this.props.bucket.id,
            first: this.state.itemCount,
          })),
        );
      };
    }
    let title = this.props.bucket.name;
    if (!this.props.bucket.public) {
      title = <div>{title}<LockIcon /></div>;
    }
    return (
      <div>
        <Dialog
          id="dialog1"
          open={this.state.deleteConfirm}
          onRequestClose={(() => this.setState(() => ({ deleteConfirm: false })))}
        >
          <DialogTitle>Are you sure?</DialogTitle>
          <DialogContent>
            <DialogContentText>
              Are you sure you want to delete the bucket forever?
            </DialogContentText>
          </DialogContent>
          <DialogActions>
            <Button id="keep" onClick={(() => this.setState(() => ({ deleteConfirm: false })))} color="primary">
              Keep it
            </Button>
            <Button id="delete2" onClick={this.deleteBucket} color="primary" autoFocus>
              Delete
            </Button>
          </DialogActions>
        </Dialog>
        <InvestBucket
          title={title}
          attributes={attributes}
          editFunc={editFunc}
          seeMoreFunc={seeMoreFunc}
          editCompositionFunc={
            this.props.bucket.isOwner ?
              (() => this.context.router.replace(`/home/composition/${this.props.bucket.id}`)) :
              null
          }
          deleteFunc={
            this.props.bucket.isOwner ?
              (() => this.setState(() => ({ deleteConfirm: true }))) :
              null
          }
          investFunc={() => this.context.router.replace(`/home/invest/${this.props.bucket.id}`)}
        />
      </div>
    );
  }
}

export default createRefetchContainer(InvestBucketRelay, {
  bucket: graphql`
    fragment InvestBucketRelay_bucket on GInvestmentBucket
    @argumentDefinitions(
      first: {type: "Int!", defaultValue: 2}
    ) {
      id
      name
      public
      isOwner
      description(first: $first) @connection(key: "InvestBucketRelay_description") {
        edges {
          node {
            id
            text
            isGood
          }
        }
        pageInfo {
          hasNextPage
        }
      }
    }
  `,
}, graphql`
    query InvestBucketRelayQuery($id: ID!, $first: Int!) {
      investBucket(idValue: $id) {
        ...InvestBucketRelay_bucket @arguments(first: $first)
      }
    }
  `);
