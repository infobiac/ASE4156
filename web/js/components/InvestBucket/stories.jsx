// @flow
import React from 'react';
import { action } from '@storybook/addon-actions';
import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import InvestBucket from './InvestBucket';
import InvestComposition from './InvestComposition';
import InvestPanel from './InvestPanel';

storiesOf('InvestBucket', module).add('Playground', () => {
  const title = text('Title of the risk section', 'Risk 1');
  const riskList = {
    good: [
      {
        id: '1',
        shortDesc: 'Good to get started',
        editMode: false,
        text: {
          onClick: action('click1'),
        },
        icon: {
          onClick: action('click3'),
        },
      }, {
        id: '2',
        shortDesc: 'Low risk',
        editMode: true,
        text: {
          onClick: action('click2'),
        },
        icon: {
          onClick: action('click3'),
        },
      },
    ],
    bad: [
      {
        id: '3',
        shortDesc: 'Low reward',
        editMode: false,
        text: {
          onClick: action('click3'),
        },
        icon: {
          onClick: action('click3'),
        },
      },
    ],
  };
  return (<InvestBucket
    title={title}
    attributes={riskList}
    editFunc={action('edit')}
    investFunc={() => {}}
  />);
});

storiesOf('InvestCompositionDontTest', module).add('Playground', () => {
  const chunks = [{
    name: 'Google',
    value: 150,
    quantity: 0.5,
    id: '1',
  }, {
    name: 'IBM',
    value: 100,
    quantity: 0.75,
    id: '2',
  }, {
    name: 'Palantir',
    value: 50,
    quantity: 2,
    id: '3',
  }, {
    name: 'Facebook',
    value: 100,
    quantity: 1,
    id: '4',
  }];
  const suggestions = [{
    name: 'GOOGL',
    value: 60,
    id: '5',
  }];
  return (
    <InvestComposition
      suggestions={suggestions}
      suggestionFieldChange={action('suggestionFieldChange')}
      total={400}
      chunks={chunks}
      chunkUpdate={action('chunkUpdate')}
      saveFunc={action('save')}
      cancelFunc={action('cancel')}
    />
  );
});
storiesOf('InvestPanelDontTest', module).add('Playground', () => (
  <InvestPanel
    bucket={{
      name: 'Some bucket',
      value: 311.45,
      ownedAmount: 30,
      history: [{ date: new Date('2017-03-03'), value: 200 }],
    }}
    available={555.55}
    investFunc={action('invest')}
    cancelFunc={action('cancel')}
  />
));
