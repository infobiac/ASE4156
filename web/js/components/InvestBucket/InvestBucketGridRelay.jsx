// @flow
import React from 'react';
import { createRefetchContainer, graphql } from 'react-relay';
import { ConnectionHandler } from 'relay-runtime';
import Grid from 'material-ui/Grid';
import Button from 'material-ui/Button';
import Card, { CardHeader, CardContent, CardActions } from 'material-ui/Card';
import InfiniteScroll from 'react-infinite-scroller';

import type { RelayContext } from 'react-relay';

import InvestBucketRelay from './InvestBucketRelay';
import EditBucket from '../EditBucket/EditBucket';
import createBucket from '../../mutations/BucketEdit/CreateBucket';

import type { InvestBucketGridRelay_profile } from './__generated__/InvestBucketGridRelay_profile.graphql';

type Props = {
  profile: InvestBucketGridRelay_profile,
  relay: RelayContext,
}
type State = {
  showDialog: bool,
  errors: ?Array<Error>,
  count: number,
}

const spacing = { xs: 12, sm: 6, md: 4 };

class InvestBucketGridRelay extends React.Component<Props, State> {
  constructor() {
    super();
    this.state = {
      showDialog: false,
      errors: [],
      count: 3,
    };
  }
  dialogAction = diagState => () => {
    this.setState(() => ({
      showDialog: diagState,
    }));
  }
  dialogSave = (name: string, publicBucket: bool, investment: number) => {
    let updater = null;
    updater = (store) => {
      const connection = ConnectionHandler.getConnection(
        store.get(this.props.profile.id),
        'InvestBucketGridRelay_investSuggestions',
      );
      const newEdge = ConnectionHandler.createEdge(
        store,
        connection,
        store.getRootField('addBucket').getLinkedRecord('bucket'),
        'GInvestmentBucketConnection',
      );
      ConnectionHandler.insertEdgeAfter(connection, newEdge);
    };
    createBucket(
      updater,
      updater,
      (response: ?Object, errors: ?[Error]) => {
        if (errors) {
          const e = errors[0];
          this.setState(() => ({ errors: [e] }));
        } else {
          this.dialogAction(false)();
        }
      },
    )(this.props.relay.environment)({ name, public: publicBucket, investment });
  }
  loadMore = () => {
    this.setState(() => ({ count: this.state.count + 3 }), () => {
      this.props.relay.refetch(() => ({ count: this.state.count }));
    });
  }
  render() {
    if (!this.props.profile.investSuggestions) {
      return null;
    }
    const emptyFillers = [];
    if (this.props.profile.investSuggestions.pageInfo.hasNextPage) {
      for (
        let i = this.props.profile.investSuggestions.edges.length;
        i < this.state.count;
        i += 1) {
        emptyFillers.push((
          <Grid item {...spacing} key={i}>
            <Card>
              <CardHeader title="Loading more buckets" />
              <CardContent>Sit tight!</CardContent>
            </Card>
          </Grid>
        ));
      }
    }
    const createMoreBucket = (
      <Card>
        <CardHeader title="Create new bucket" />
        <CardContent>Here you can add a new bucket!</CardContent>
        <CardActions>
          <Button dense color="primary" aria-label="add" onClick={this.dialogAction(true)}>
              New
          </Button>
        </CardActions>
      </Card>);
    return (
      <InfiniteScroll
        loadMore={this.loadMore}
        hasMore={
          this.props.profile.investSuggestions
          && this.props.profile.investSuggestions.pageInfo.hasNextPage
        }
      >
        <Grid container spacing={16} align="stretch">
          <Grid item {...spacing}>
            {createMoreBucket}
          </Grid>
          {
            this.props.profile.investSuggestions ?
              this.props.profile.investSuggestions.edges.map(b => (b && b.node ? (
                <Grid item {...spacing} key={b.node.id}>
                  <InvestBucketRelay profile={this.props.profile} bucket={b.node} />
                </Grid>
              ) : null)) : null
          }
          {emptyFillers}
          <Grid item {...spacing}>
            {
              this.props.profile.investSuggestions
              && this.props.profile.investSuggestions.pageInfo.hasNextPage ?
                <div onChange={v => v && this.loadMore()} />
              : createMoreBucket
            }
          </Grid>
          {
            this.state.showDialog ?
              <EditBucket
                save={this.dialogSave}
                cancel={this.dialogAction(false)}
                errors={this.state.errors}
              /> :
            null
          }
        </Grid>
      </InfiniteScroll>
    );
  }
}

export default createRefetchContainer(InvestBucketGridRelay, {
  profile: graphql`
    fragment InvestBucketGridRelay_profile on GProfile
    @argumentDefinitions(
      count: {type: "Int!", defaultValue: 3}
    ) {
      id
      investSuggestions(first: $count) @connection(key: "InvestBucketGridRelay_investSuggestions") {
        edges {
          node {
            id
            ...InvestBucketRelay_bucket
          }
        }
        pageInfo {
          hasNextPage
        }
      }
    }
  `,
}, graphql`
  query InvestBucketGridRelayQuery($count: Int!) {
    viewer {
      profile {
        ...InvestBucketGridRelay_profile @arguments(count: $count)
      }
    }
  }
`);
