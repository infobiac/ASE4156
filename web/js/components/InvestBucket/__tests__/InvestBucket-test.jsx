import React from 'react';
import { shallow } from 'enzyme';
import InvestBucket from '../InvestBucket';

describe('InvestBucket', () => {
  it('exists', () => {
    expect(InvestBucket).toBeTruthy();
  });

  it('renders', () => {
    const comp = shallow(<InvestBucket
      title="Random"
      attributes={{ good: [], bad: [] }}
      investFunc={jest.fn()}
    />);
    expect(comp).toMatchSnapshot();
  });

  it('displays good attributes', () => {
    const comp = shallow(<InvestBucket
      title="Random"
      attributes={{ good: [{ id: 1, shortDesc: 'Attr' }], bad: [] }}
      investFunc={jest.fn()}
    />);
    expect(comp).toMatchSnapshot();
  });

  it('displays bad attributes', () => {
    const comp = shallow(<InvestBucket
      title="Random"
      attributes={{ good: [], bad: [{ id: 1, shortDesc: 'Attr' }] }}
      investFunc={jest.fn()}
    />);
    expect(comp).toMatchSnapshot();
  });

  it('edit attribute', () => {
    const editFunc = jest.fn();
    const comp = shallow(<InvestBucket
      title="Random"
      attributes={{
        good: [{
            id: 1,
            shortDesc: 'Attr',
            editMode: true,
          text: { onChange: editFunc, id: 'editTxt' },
        }],
          bad: [],
      }}
      investFunc={jest.fn()}
    />);
    const obj = { target: { value: '3' } };
    expect(editFunc.mock.calls.length).toEqual(0);
    comp.find('#editTxt').simulate('change', obj);
    expect(editFunc.mock.calls.length).toEqual(1);
    expect(editFunc.mock.calls[0][0]).toEqual(obj);
  });

  it('add attribute', () => {
    const editFunc = jest.fn();
    const comp = shallow(<InvestBucket
      title="Random"
      attributes={{
        good: [],
          bad: [],
      }}
      editFunc={editFunc}
      investFunc={jest.fn()}
    />);
    comp.find('#launch-edit').simulate('click');
    expect(comp).toMatchSnapshot();
    comp.find('#name').simulate('change', { target: { value: 'some text' } });
    expect(comp).toMatchSnapshot();
    comp.find('#good-or-bad').simulate('click');
    expect(comp).toMatchSnapshot();
    comp.find('#name').simulate('keyPress', { charCode: 10 });
    expect(editFunc.mock.calls.length).toEqual(0);
    comp.find('#name').simulate('keyPress', { charCode: 13 });
    expect(editFunc.mock.calls.length).toEqual(1);
    expect(editFunc.mock.calls[0][0]).toEqual('some text');
    expect(editFunc.mock.calls[0][1]).toEqual(false);
    expect(comp).toMatchSnapshot();
  });

  it('see more', () => {
    const seeMoreFunc = jest.fn();
    const comp = shallow(<InvestBucket
      title="Random"
      attributes={{
        good: [],
          bad: [],
      }}
      seeMoreFunc={seeMoreFunc}
      investFunc={jest.fn()}
    />);
    expect(seeMoreFunc.mock.calls.length).toEqual(0);
    comp.find('#more').simulate('click');
    expect(seeMoreFunc.mock.calls.length).toEqual(1);
  });

  it('delete', () => {
    const deleteFunc = jest.fn();
    const comp = shallow(<InvestBucket
      title="Random"
      attributes={{
        good: [],
          bad: [],
      }}
      deleteFunc={deleteFunc}
      investFunc={jest.fn()}
    />);
    expect(deleteFunc.mock.calls.length).toEqual(0);
    comp.find('#delete').simulate('click');
    expect(deleteFunc.mock.calls.length).toEqual(1);
  });

  it('edit composition', () => {
    const editCompositionFunc = jest.fn();
    const comp = shallow(<InvestBucket
      title="Random"
      attributes={{
        good: [],
          bad: [],
      }}
      editCompositionFunc={editCompositionFunc}
      investFunc={jest.fn()}
    />);
    expect(editCompositionFunc.mock.calls.length).toEqual(0);
    comp.find('#edit-comp').simulate('click');
    expect(editCompositionFunc.mock.calls.length).toEqual(1);
  });
});

/*
type ItemObj = {
  id: string,
  shortDesc: string,
  icon?: any,
  text?: any,
  editMode: bool,
}
type Props = {
  title: Node,
  attributes: {
    good: Array<ItemObj>,
    bad: Array<ItemObj>,
  },
  editFunc?: ?(text: string, isGood: bool) => void,
  seeMoreFunc?: ?() => void,
  editCompositionFunc?: ?() => void,
  deleteFunc?: ?() => void,
  investFunc: () => void,
}
 */
