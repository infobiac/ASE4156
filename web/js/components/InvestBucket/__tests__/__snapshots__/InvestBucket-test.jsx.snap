// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvestBucket add attribute 1`] = `
<Card
  raised={false}
>
  <withStyles(CardHeader)
    title="Random"
  />
  <withStyles(CardContent)>
    <withStyles(List)>
      <withStyles(ListItem)>
        <withStyles(ListItemIcon)
          id="good-or-bad"
          onClick={[Function]}
        >
          <pure(TrendingUp) />
        </withStyles(ListItemIcon)>
        <TextField
          autoFocus={true}
          fullWidth={true}
          id="name"
          label="New Attribute"
          margin="dense"
          onChange={[Function]}
          onKeyPress={[Function]}
          required={false}
          select={false}
          type="text"
          value=""
        />
      </withStyles(ListItem)>
    </withStyles(List)>
  </withStyles(CardContent)>
  <withStyles(CardActions)>
    <withStyles(Button)
      color="primary"
      dense={true}
      id="invest"
      onClick={[Function]}
    >
      Invest
    </withStyles(Button)>
  </withStyles(CardActions)>
</Card>
`;

exports[`InvestBucket add attribute 2`] = `
<Card
  raised={false}
>
  <withStyles(CardHeader)
    title="Random"
  />
  <withStyles(CardContent)>
    <withStyles(List)>
      <withStyles(ListItem)>
        <withStyles(ListItemIcon)
          id="good-or-bad"
          onClick={[Function]}
        >
          <pure(TrendingUp) />
        </withStyles(ListItemIcon)>
        <TextField
          autoFocus={true}
          fullWidth={true}
          id="name"
          label="New Attribute"
          margin="dense"
          onChange={[Function]}
          onKeyPress={[Function]}
          required={false}
          select={false}
          type="text"
          value="some text"
        />
      </withStyles(ListItem)>
    </withStyles(List)>
  </withStyles(CardContent)>
  <withStyles(CardActions)>
    <withStyles(Button)
      color="primary"
      dense={true}
      id="invest"
      onClick={[Function]}
    >
      Invest
    </withStyles(Button)>
  </withStyles(CardActions)>
</Card>
`;

exports[`InvestBucket add attribute 3`] = `
<Card
  raised={false}
>
  <withStyles(CardHeader)
    title="Random"
  />
  <withStyles(CardContent)>
    <withStyles(List)>
      <withStyles(ListItem)>
        <withStyles(ListItemIcon)
          id="good-or-bad"
          onClick={[Function]}
        >
          <pure(TrendingDown) />
        </withStyles(ListItemIcon)>
        <TextField
          autoFocus={true}
          fullWidth={true}
          id="name"
          label="New Attribute"
          margin="dense"
          onChange={[Function]}
          onKeyPress={[Function]}
          required={false}
          select={false}
          type="text"
          value="some text"
        />
      </withStyles(ListItem)>
    </withStyles(List)>
  </withStyles(CardContent)>
  <withStyles(CardActions)>
    <withStyles(Button)
      color="primary"
      dense={true}
      id="invest"
      onClick={[Function]}
    >
      Invest
    </withStyles(Button)>
  </withStyles(CardActions)>
</Card>
`;

exports[`InvestBucket add attribute 4`] = `
<Card
  raised={false}
>
  <withStyles(CardHeader)
    title="Random"
  />
  <withStyles(CardContent)>
    <withStyles(List) />
    <withStyles(Button)
      aria-label="add"
      color="primary"
      fab={true}
      id="launch-edit"
      onClick={[Function]}
    >
      <pure(Add) />
    </withStyles(Button)>
  </withStyles(CardContent)>
  <withStyles(CardActions)>
    <withStyles(Button)
      color="primary"
      dense={true}
      id="invest"
      onClick={[Function]}
    >
      Invest
    </withStyles(Button)>
  </withStyles(CardActions)>
</Card>
`;

exports[`InvestBucket displays bad attributes 1`] = `
<Card
  raised={false}
>
  <withStyles(CardHeader)
    title="Random"
  />
  <withStyles(CardContent)>
    <withStyles(List)>
      <withStyles(ListItem)
        key="1"
      >
        <withStyles(ListItemIcon)>
          <pure(TrendingDown) />
        </withStyles(ListItemIcon)>
        <withStyles(ListItemText)
          autoFocus={true}
          id="attr"
          key="1"
          primary="Attr"
        />
      </withStyles(ListItem)>
    </withStyles(List)>
  </withStyles(CardContent)>
  <withStyles(CardActions)>
    <withStyles(Button)
      color="primary"
      dense={true}
      id="invest"
      onClick={[Function]}
    >
      Invest
    </withStyles(Button)>
  </withStyles(CardActions)>
</Card>
`;

exports[`InvestBucket displays good attributes 1`] = `
<Card
  raised={false}
>
  <withStyles(CardHeader)
    title="Random"
  />
  <withStyles(CardContent)>
    <withStyles(List)>
      <withStyles(ListItem)
        key="1"
      >
        <withStyles(ListItemIcon)>
          <pure(TrendingUp) />
        </withStyles(ListItemIcon)>
        <withStyles(ListItemText)
          autoFocus={true}
          id="attr"
          key="1"
          primary="Attr"
        />
      </withStyles(ListItem)>
    </withStyles(List)>
  </withStyles(CardContent)>
  <withStyles(CardActions)>
    <withStyles(Button)
      color="primary"
      dense={true}
      id="invest"
      onClick={[Function]}
    >
      Invest
    </withStyles(Button)>
  </withStyles(CardActions)>
</Card>
`;

exports[`InvestBucket renders 1`] = `
<Card
  raised={false}
>
  <withStyles(CardHeader)
    title="Random"
  />
  <withStyles(CardContent)>
    <withStyles(List) />
  </withStyles(CardContent)>
  <withStyles(CardActions)>
    <withStyles(Button)
      color="primary"
      dense={true}
      id="invest"
      onClick={[Function]}
    >
      Invest
    </withStyles(Button)>
  </withStyles(CardActions)>
</Card>
`;
